lightningd/htlc_end.h

```diff
@@ -158,6 +164,7 @@ struct htlc_in *new_htlc_in(const tal_t *ctx,
 			    const struct secret *shared_secret TAKES,
 			    const struct pubkey *path_key TAKES,
 			    const u8 *onion_routing_packet,
+			    const struct tlv_field *extra_tlvs TAKES,
```

Comment:
Same comment regarding `TAKES`: you probably don't want this to reparent the `tlv_field` array, because you then iterate and clone the values over, leaving the old values potentially dangling. Remove the `TAKES` and don't call with `take(extra_tlvs)` and it should work as expected (`valgrind` is likely not triggering a memleak because you don't call with `take()` which'd trigger the reparenting).

Reply:
Good catch.  I created a helper to dup this array properly...

Reply:
Yes, actually `valgrind` complained for the parts where the caller actually used `take()` for the `extra_tlvs`, but, iirc because the `path_key` variable takes from the same parent (or is a pointer to one of the fields?). I just forgot to remove it from all the places. It might make sense to check with `valgrind` again with an actuall caller using `take()`.